### PR TITLE
Handle full page session refreshes

### PR DESCRIPTION
**Pull request by alsocalledchris**
Updated to better handle full page refreshes - remove session URL from redirect URL and URL decode absolute URL so don't get errors on redirect.

### CODE REVIEW

File: libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/SFHybridViewController.m [hunk 1]
```diff
@@ -425,6 +425,15 @@ - (NSURL *)frontDoorUrlWithReturnUrl:(NSString *)returnUrlString returnUrlIsEnco
     NSURL *instUrl = creds.apiUrl;
     NSString *fullReturnUrlString = returnUrlString;
     
+    // Catch if we have a absoute URL, if we do then url decode to avoid
+    // errors with 'Invalid page redirection'
+    if(createAbsUrl && [returnUrlString hasPrefix:@"http"]) {
+        NSString *decodedUrlString = [(NSString *)returnUrlString stringByReplacingOccurrencesOfString:@"+" withString:@" "];
```
Comment (bhariharan) [inline libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/SFHybridViewController.m]: I don't think you need the explicit cast here - `returnUrlString` is of type `NSString *`.

File: libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/SFHybridViewController.m [hunk 2]
```diff
@@ -425,6 +425,15 @@ - (NSURL *)frontDoorUrlWithReturnUrl:(NSString *)returnUrlString returnUrlIsEnco
     NSURL *instUrl = creds.apiUrl;
     NSString *fullReturnUrlString = returnUrlString;
     
+    // Catch if we have a absoute URL, if we do then url decode to avoid
+    // errors with 'Invalid page redirection'
+    if(createAbsUrl && [returnUrlString hasPrefix:@"http"]) {
```
Comment (bhariharan) [inline libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/SFHybridViewController.m]: Shouldn't this be `https`? I don't think any pages within the `force.com` domain support `http`.

File: libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/SFHybridViewController.m [hunk 3]
```diff
@@ -591,8 +600,10 @@ - (void) webView:(WKWebView *) webView decidePolicyForNavigationAction:(WKNaviga
             [self refreshCredentials:[SFUserAccountManager sharedInstance].currentUser.credentials
              completion:^(SFOAuthInfo *authInfo, SFUserAccount *userAccount) {
                  [SFUserAccountManager sharedInstance].currentUser = userAccount;
+                 //Remove the session URL from the return URL as we have taken over the refresh process
+                 NSString *sessionRemovedRefreshUrl = [refreshUrl stringByReplacingOccurrencesOfString:@"%2Fvisualforce%2Fsession%3Furl%3D" withString:@""];
```
Comment (bhariharan) [inline libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/SFHybridViewController.m]: Where is this `visualforce` URL coming from?

**Comment by brandonpage** [comment]
## Danger has errored
[!] Invalid `Dangerfile-Lib.rb` file: [31mWorkspace file not found at path `/Users/iosadmin/03b72f19/workspace/SalesforceMobileSDK.xcworkspace`[0m
```
 #  from .circleci/Dangerfile-Lib.rb:16
 #  -------------------------------------------
 #  if ENV.has_key?('JENKINS_URL') && File.file?(coverage_results)
 >    xcov.report(
 #        scheme: 'UnitTests',
 #  -------------------------------------------
```

<p align="right" data-meta="generated_by_CI-Result-SalesforceHybridSDK">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/SFHybridViewController.m (hunk 2 above)
Comment (chriswellingsfairsailcom) [reply libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/SFHybridViewController.m]: It was just the same as the line below - will catch both https and http

File: libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/SFHybridViewController.m (hunk 1 above)
Comment (chriswellingsfairsailcom) [reply libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/SFHybridViewController.m]: Ah OK, thanks

File: libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/SFHybridViewController.m (hunk 3 above)
Comment (chriswellingsfairsailcom) [reply libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/SFHybridViewController.m]: No idea - it is sometimes (mostly) there and others not (when not the refresh worked before this pull request's change) - so is causes the issue.

**Comment by chriswellingsfairsailcom** [comment]
@bhariharan not sure how Android is working without this extra code - but it probably needs it too if you agree the above solution is correct.

**Comment by bhariharan** [comment]
@alsocalledchris I think [this PR](https://github.com/forcedotcom/SalesforceMobileSDK-iOS/pull/2438) should fix this.

**Comment by alsocalledchris** [comment]
Many thanks @bhariharan, and thanks for picking up the other issues I have logged. When will those fixes make it into a release? :)

**Comment by bhariharan** [comment]
I will post updates on the issues filed as and when I make progress.

**Pull request closed without merging**